### PR TITLE
Refactor pwnedPassword to use hash range API

### LINKED ISSUE #22: Refactor pwnedPassword.js to use hash range API
Depending on my availability over the next bit, I would be interested in trying to implement this feature.

Currently the pwnedPassword check uses the v1 api, but it would be nice to use the more secure v2 hash range api.  This module could depend upon `pwnedPasswordRange.js` and provide the higher level API without needing to search through the results.

I was envisioning using this as a type of password strength tester for user signup beyond the basic length/character set, so there could be a case where you were ok with 5 pwns but not 5,000.  I would add a `maxPwns: number` parameter to the options object (convenient it's already there!) to facilitate that behavior.  Default `maxPwns` would be 1, and the method would return a boolean as it does currently.

### PR DESCRIPTION
**Pull request by danieladams456**
PR for feature enhancement #22 
This is my first time contributing to an open source library, so we'll see how it goes. 😄 

Closes #22

### CODE REVIEW

**Comment by wKovacs64** [comment]
> This is my first time contributing to an open source library, so we'll see how it goes.

✨ Hooray, that's great! ✨

I'm happy to help, so feel free to ask any questions here in the PR if needed.

**Comment by danieladams456** [comment]
What is your local development toolchain? I did a work project with Angular CLI and typescript but haven't used babel before. Ideally I'd like to have VSCode open with two ES6 files: the library file I'm working on and a small test script. Each time I save either one the test script gets run and output printed to the console. Could I do that with babel-node?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by wKovacs64** [comment]
On mobile atm, can elaborate later if needed, but `yarn test:watch` (or `npm run test:watch`) should do it.

**Comment by danieladams456** [comment]
OK I'll look at what that's referencing. My first shot at installing the sha1 package I did with npm and was confused by why package-lock.json wasn't tracked, but then I saw the yarn lock file so I redid with yarn.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by wKovacs64** [comment]
Yes, I prefer `yarn` to `npm` at the moment (subject to change) so I track `yarn.lock` and ignore `package-lock.json`.

You probably figured it all out already, but the `test:watch` script in `package.json` runs the tests in watch mode which will rerun upon detecting changes. I use [Jest](https://facebook.github.io/jest/), which serves as the testing framework, test runner, assertion library, mocking library, etc. (I normally prefer the Unix philosophy of small modules doing one thing well, but Jest is an exception.) Since we also have `babel-jest` installed (in `devDependencies`), Jest will automatically run everything through Babel, allowing you to use ES2015/ES6/etc.

After running `yarn test:watch` in the terminal, you could then create `foo.test.js` and Jest will only run that test suite since it's new/changed. Creating and modifying `foo.js` will also trigger test reruns.

**Comment by danieladams456** [comment]
Thanks for the tip on running new tests only.  One more preliminary question around testing:  I know unit tests are only supposed to go one function deep, but in this case it almost seems to make more sense to not mock the `pwnedPasswordRange` but rather the API response?  That way you get more of a integration test but done statically?  Since the code path is so small, it shouldn't be too confusing if a change to `pwnedPasswordRange` also causes `pwnedPassword` to fail.  That  would probably be better than forgetting to change the mock of `pwnedPasswordRange` upon function change.

Even though this is a super simple change, my goal for getting this done is by the end of the week.  My wife and I have a 3 week old so there's not a ton of free time. 😄

**Comment by wKovacs64** [comment]
> Thanks for the tip on running new tests only. One more preliminary question around testing: I know unit tests are only supposed to go one function deep, but in this case it almost seems to make more sense to not mock the pwnedPasswordRange but rather the API response? That way you get more of a integration test but done statically? Since the code path is so small, it shouldn't be too confusing if a change to pwnedPasswordRange also causes pwnedPassword to fail. That would probably be better than forgetting to change the mock of pwnedPasswordRange upon function change.

Yeah, totally agree.

As remote API responses are already mocked for the `/range` endpoint, you shouldn't even need to add any new mocking as long as the hash of the password you use starts with the same characters that make up `RANGE_VALID` in the existing test fixtures. Maybe that just means changing the [`PASSWORD_PWNED`](https://github.com/wKovacs64/hibp/blob/d6eed3aad5f15b69352ecbe1993f0652db6fe3b6/test/fixtures.js#L18) value to `P@ssw0rd` or something. (I used the same hash prefix used in the [example](https://api.pwnedpasswords.com/range/21BD1) from the API documentation and the hash of `P@ssw0rd` is one of the hashes in the resulting range.)

> Even though this is a super simple change, my goal for getting this done is by the end of the week. My wife and I have a 3 week old so there's not a ton of free time.

No sweat, and congratulations 🎊 . Mine are a bit older now, but I remember those days. 😴

**Comment by danieladams456** [comment]
Sounds good. I'll use that one for the found case and write up a quick script to generate me another that starts with the same hash but is not contained in the API response.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by wKovacs64** [comment]
Hmm, we might need to tweak the mocked API routes and responses after all. Not sure there is a way to currently distinguish between found and clean with regards to searching by range.

**Comment by danieladams456** [comment]
Another thing I was wondering is whether in the same major version upgrade it would be worth changing the `pwnedPasswordRange` call to return an object like:
```js
[
  {
    suffix: '1826128412394873'
    count: 2
  },
  {
    suffix: '1826128418461348'
    count: 16
  }
]
```

Then the password call would be able to do:
```js
res.filter(item => item.suffix === searchSuffix).map(item => item.count)
```
The `.includes` [here](https://github.com/wKovacs64/hibp/blob/develop/src/pwnedPasswordRange.js#L22) is a nice quick way of checking existence, but if you want the higher level API you should use `pwnedPassword` instead.

**Comment by wKovacs64** [comment]
I briefly wrestled with that idea during the initial implementation. The API returns JSON data for all of the other endpoints, and I've essentially been passing that data back through to the library consumer untouched. In this case, the endpoint returns plain text (an unfortunate choice on the part of the remote API, IMHO), so the decision was whether to forward along the data as-is, just like the other modules do, or to become opinionated about how it should look and convert it to JSON first. I erred on the side of consistency with regards to being unopinionated and left it alone (as much as I would prefer it to be JSON). Although one could argue returning JSON like the other modules would be more consistent in a different way, so...

I wonder if there is any use case where a consumer would want/need to access the plain text? 🤔  If not, I could possibly be convinced to transform it to JSON (in a separate PR).

**Comment by danieladams456** [comment]
From his [blog post](https://www.troyhunt.com/ive-just-launched-pwned-passwords-version-2):
>For example, the result of that range query is not a neatly formatted piece of JSON, it's just colon delimited rows. That impacts my ability to add attributes at a later date and pretty much locks in the current version to today's behaviour, but it saves on the response size. Yes, I know some curly braces and quotes wouldn't add a lot of size, but every byte counts when volumes get large.

I will put the parsing logic in this method for now, then later before release if you want in another PR you could refactor to move that down to the `pwnedPasswordRange` function.

**Comment by wKovacs64** [comment]
You know what's sad? I hadn't even seen that blog post. 😞 

> I will put the parsing logic in this method for now, then later before release if you want in another PR you could refactor to move that down to the pwnedPasswordRange function.

Sounds good, we'll revisit later then.

**Comment by danieladams456** [comment]
I think it's ready for review! 😃 Most of the commits had to do with expanding the test mocking to reflect the API enough to do the parsing/filtering.

One thing I do like about RxJS Observables is that you can do the `.map()` and `.filter()` directly on the Observable.  I tried to adapt the pattern to work inside a `.then()`.  I think it turned out [pretty clean](https://github.com/danieladams456/hibp/blob/cf5ea291d62db9ea4ccea4b40ccdfc7da676c869/src/pwnedPassword.js#L36-L48).

File: test/fixtures.js [hunk 1]
```diff
@@ -16,14 +16,19 @@ export const EMAIL_PASTED = 'foo789@bar.com';
...
-export const OPTS_ISAHASH = { isAHash: true };
-export const RANGE_VALID = '21BD1';
+export const RANGE_PASSWORD_PWNED = '5BAA6';
+export const RANGE_PASSWORD_CLEAN = '01EEC';
 export const RANGE_INVALID = '21BD';
-export const RESPONSE_STR = 'Stuff';
 export const RESPONSE_OBJ = {};
 export const RESPONSE_ARY = [];
 export const RESPONSE_CLEAN = null;
+export const RANGE_PASSWORD_PWNED_RESPONSE = `003D68EB55068C33ACE09247EE4C639306B:3
```
Comment (wKovacs64) [inline test/fixtures.js]: Let's add `common-tags` as a dev dependency and format these with `stripIndents` like in [`pwned`](https://github.com/wKovacs64/pwned/blob/6970c5911e6dda356c806fd03ddb1f77b491316c/test/fixtures.js).

File: src/pwnedPassword.js [hunk 2]
```diff
@@ -1,48 +1,20 @@
...
 /**
  * Fetches the pwned status for the given password, indicating whether or not it
- * has been previously exposed in a breach. Passwords can be plain text or a
- * SHA1 hash. The remote API will automatically attempt to discern between the
- * two, but in the case where the password you wish to check is actually a
- * plain text string containing a hash and you don't want the API to treat it as
- * a hash, you can override the auto detection behavior by setting the isAHash
- * option to true.
+ * has been previously exposed in a breach. The password is given in plaintext,
+ * but only the first 5 characters of its SHA1 hash will be sumitted to the API.
```
Comment (wKovacs64) [inline src/pwnedPassword.js]: `sumitted` should be `submitted`

File: src/pwnedPassword.js [hunk 3]
```diff
@@ -1,48 +1,20 @@
...
- * @param {string} password a password (plain text string or SHA1 hash)
- * @param {Object} [options] a configuration object
- * @param {boolean} [options.isAHash] the pre-hashed password is a hash
- * (default: false)
- * @returns {Promise} a Promise which resolves to true if the given password has
- * been exposed in a breach (or false if not), or rejects with an Error
+ * @param {string} password a password in plaintext
+ * @returns {Promise} a Promise which resolves to the number of times the password has
+ * been breached, or rejects with an Error
  * @example
```
Comment (wKovacs64) [inline src/pwnedPassword.js]: Can we change the `@returns` verbiage here to `a Promise which resolves to the number of times the password has been exposed in a breach, or rejects with an Error`?

File: test/fixtures.js (hunk 1 above)
Comment (danieladams456) [reply test/fixtures.js]: Very cool!

**Comment by wKovacs64** [comment]
LGTM! 👍 Thank you!

**Pull request merged**